src/ScheduleFunctions.cpp

```diff
@@ -1138,149 +949,210 @@ class ShiftLoopNest : public IRMutator2 {
     }
 
 public:
-    ShiftLoopNest(const map<string, Expr> &s) : shifts(s) {}
+    explicit ShiftLoopNest(const map<string, Expr> &s) : shifts(s) {}
+
+    template<typename T>
+    static T apply_shift(const map<string, Expr> &shifts, const T &node) {
+        if (shifts.empty()) {
+            return node;
+        }
+        ShiftLoopNest visitor(shifts);
+        return visitor.mutate(node);
+    }
 };
 
-// Inject the allocation and realization of a group of functions which are
-// to be fused into an existing loop nest using its schedule.
-class InjectGroupRealization : public IRMutator2 {
+class InjectFunctionRealization : public IRMutator2 {
 public:
-    // Member of the fused loop starting from the first to be realized to the last.
-    const vector<Function> &group;
-    // List of booleans indicating if group[i] is an output
-    const vector<bool> &is_output_list;
-    bool found_store_level, found_compute_level;
-    const Target &target;
-    const map<string, Function> &env;
-    LoopLevel compute_level;
-    LoopLevel store_level;
+    InjectFunctionRealization(const vector<Function> &funcs, const vector<bool> &is_output_list, const Target &target,
+                              const map<string, Function> &env)
+            : funcs(funcs)
+            , is_output_list(is_output_list)
```

Comment:
non-idiomatic. Check the style used elsewhere for multi-line initializers. I think we put commas on the end of the previous line.

Reply:
@abadams  -- do we prefer A:
```
    InjectFunctionRealization(const vector<Function> &funcs,
                              const vector<bool> &is_output_list,
                              const Target &target,
                              const map<string, Function> &env)
        : funcs(funcs),
          is_output_list(is_output_list),
          target(target),
          env(env),
          compute_level(funcs[0].schedule().compute_level()),
          store_level(funcs[0].schedule().store_level()) {}
```
or B:
```
InjectFunctionRealization(const vector<Function> &funcs,
                              const vector<bool> &is_output_list,
                              const Target &target,
                              const map<string, Function> &env) : funcs(funcs),
                                                                  is_output_list(is_output_list),
                                                                  target(target),
                                                                  env(env),
                                                                  compute_level(funcs[0].schedule().compute_level()),
                                                                  store_level(funcs[0].schedule().store_level()) {}
```